crates/oxc_semantic/src/ast_types_bitset.rs

```diff
@@ -0,0 +1,112 @@
+use oxc_ast::{AstType, ast_kind::AST_TYPE_MAX};
+
+const USIZE_BITS: usize = usize::BITS as usize;
+
+/// Number of bytes required for bit set which can represent all [`AstType`]s.
+// Need to add plus one here because 0 is a possible value, but requires at least one bit to represent it.
+const NUM_USIZES: usize = ((AST_TYPE_MAX + 1) as usize).div_ceil(USIZE_BITS);
+
+/// Bit set with a bit for each [`AstType`].
+#[derive(Debug, Clone)]
+pub struct AstTypesBitset([usize; NUM_USIZES]);
+
+impl AstTypesBitset {
+    /// Create empty [`AstTypesBitset`] with no bits set.
+    pub const fn new() -> Self {
+        Self([0; NUM_USIZES])
+    }
+
+    /// Returns `true` if bit is set for provided [`AstType`].
+    pub const fn has(&self, ty: AstType) -> bool {
+        let (index, mask) = Self::index_and_mask(ty);
+        (self.0[index] & mask) != 0
+    }
+
+    /// Set bit for provided [`AstType`].
+    pub const fn set(&mut self, ty: AstType) {
+        let (index, mask) = Self::index_and_mask(ty);
+        self.0[index] |= mask;
+    }
+
+    /// Returns `true` if any bit is set in both `self` and `other`.
+    pub fn intersects(&self, other: &Self) -> bool {
+        let mut intersection = 0;
+        for (&a, &b) in self.0.iter().zip(other.0.iter()) {
+            intersection |= a & b;
+        }
+        intersection != 0
+    }
+
+    /// Returns `true` if all bits in `other` are set in `self`.
+    pub fn contains(&self, other: &Self) -> bool {
+        for (&a, &b) in self.0.iter().zip(other.0.iter()) {
+            if a & b != b {
+                return false;
+            }
+        }
+        true
+    }
```

Comment:
Probably preferable to make this branchless:

```rs
pub fn contains(&self, other: &Self) -> bool {
    let mut mismatches = 0;
    for (&a, &b) in self.0.iter().zip(other.0.iter()) {
        let set_in_both = a & b;
        // 0 if `set_in_both == b`
        let mismatch = set_in_both ^ b;
        mismatches |= mismatch;
    }
    mismatches == 0
}
```

https://godbolt.org/z/19sv53KMM